Updated the .NET Core to use the new SQLClient lib

cmatskas opened the pull request:
The doc currently references an outdated library. The recommended changes use the SQLClient library with Azure.Identity to acquire tokens for the DBContext in every environment with no other code or config changes

Commit:
Updated the .NET Core to use the new SQLClient lib

The doc currently references an outdated library. The recommended changes use the SQLClient library with Azure.Identity to acquire tokens for the DBContext in every environment with no other code or config changes

mderriey commented on articles/app-service/app-service-web-tutorial-connect-msi.md:
````diff
@@ -136,29 +136,60 @@ Type `Ctrl+F5` to run the app again. The same CRUD app in your browser is now co
 
 ```json
-"Server=tcp:<server-name>.database.windows.net,1433;Database=<database-name>;"
+"Server=tcp:<server-name>.database.windows.net;Authentication=Active Directory.Device Code Flow; Database=<database-name>;"
````
My understanding here is that the value for the `Authentication` keyword is not important, but needs to match the one from our custom `SqlAuthenticationProvider`.

Is that correct?
If it is, then I'd suggest we either explain why we picked the "device code flow" authentication type.

cephalin replied:
I agree. I have the same question about the type choice (and looks like there's a . in there that shouldn't be there according to the AAD documentation). And if it's not important, then why the need to specify at all?

cmatskas replied:
@mderriey @cephalin added a note that explains why AuthCode Flow was chosen here

mderriey commented on articles/app-service/app-service-web-tutorial-connect-msi.md:
```diff
@@ -136,29 +136,60 @@ Type `Ctrl+F5` to run the app again. The same CRUD app in your browser is now co
+services.AddControllersWithViews();
+services.AddDbContext<MyDatabaseContext>(options =>
+{
+    SqlAuthenticationProvider.SetProvider(
+        SqlAuthenticationMethod.ActiveDirectoryDeviceCodeFlow, 
+        new CustomAzureSQLAuthProvider());
```
nit: this lambda is executed on every single request, so I'd move this outside given it only needs to be configured once.

cephalin replied:
If it's a nit, I vote for simplicity on part of the reader instead of "complexity in favor of best practice". At the end of the day this doc's priority is demonstrating the connectivity scenario and we want to simplify the code where it's possible/appropriate.

cmatskas replied:
Thanks both. I agree that this is not recommended practice but we should strive to keep the sample simple. We should add a note below that actually informs the reader that this method get's called with every call and therefore should be move outside. Thoughts @mderriey ?

mderriey commented on articles/app-service/app-service-web-tutorial-connect-msi.md:
```diff
@@ -136,29 +136,60 @@ Type `Ctrl+F5` to run the app again. The same CRUD app in your browser is now co
 
-> [!TIP]
-> If the Azure AD user you configured has access to multiple tenants, call `GetAccessTokenAsync("https://database.windows.net/", tenantid)` with the desired tenant ID to retrieve the proper access token.
+That's every thing you need to connect to SQL Database. When debugging in Visual Studio, your code uses the Azure AD user you configured in [Set up Visual Studio](#set-up-visual-studio). You'll set up SQL Database later to allow connection from the managed identity of your App Service app. The `DefaultAzureCredential` class caches the token in memory and retrieves it from Azure AD just before exp ...
```
```suggestion
That's everything you need to connect to SQL Database. When debugging in Visual Studio, your code uses the Azure AD user you configured in [Set up Visual Studio](#set-up-visual-studio). You'll set up SQL Database later to allow connection from the managed identity of your App Service app. The `DefaultAzureCredential` class caches the token in memory and retrieves it from Azure AD just before expiration. You don't need any custom code to refresh the token.
```

> The `DefaultAzureCredential` class caches the token in memory and retrieves it from Azure AD just before expiration. You don't need any custom code to refresh the token.

I'm fairly sure that's not the case, unless something changed recently.
My understanding is that `Azure.Identity` itself does not handle token caching / renewal; the Azure client libraries like Blobs, Queues, ServiceBus, etc... do this through a custom HTTP pipeline.

Happy to be proven wrong, though 😆 &mdash; @schaabs?

cephalin replied:
> > The `DefaultAzureCredential` class caches the token in memory and retrieves it from Azure AD just before expiration. You don't need any custom code to refresh the token.
> 
> I'm fairly sure that's not the case, unless something changed recently.
> My understanding is that `Azure.Identity` itself does not handle token caching / renewal; the Azure client libraries like Blobs, Queues, ServiceBus, etc... do this through a custom HTTP pipeline.
> 
> Happy to be proven wrong, though 😆 — @schaabs?

In that case, I question the appropriateness of the change at all. The AppAuthentication approach so far is actually demonstrably simpler, especially since previously it was just 2 lines of c# code with no need for custom providers, AND it caches tokens. Now we're telling the customers now they need to write more code AND they need to add code to cache tokens? I don't mean to be harsh but this isn't an attractive scenario.

schaabs replied:
I agree that having the user code have to cache the token is less than ideal, but not all credential types in Azure.Identity cache tokens. We hope to provide a better story around SQL authentication in the long term, most likely by providing a `SqlAuthenticationProvider` similar to what @cmatskas has provided in this sample, which would handle any caching for the user, but while that is on our roadmap we currently don't have the work planned.

That said I don't think you can compare this new sample using Azure.Identity, with the previous sample using the AppAuthentication library 1:1. The old sample simply acquired a token and statically set that token on the connection. So it would seem to me that the old sample wouldn't handle token expiration at all. But perhaps I'm wrong here, as I'm not totally familiar with how EF creates creates and reuses connections.

cephalin replied:
@schaabs Sorry I had a long break in April. Can we compromise by going ahead with the change on my side but putting `SqlAuthenticationProvider` on the roadmap on your end?

I agree about not making 1:1 comparison with AppAuthentication. However I did read docs somewhere that said that AppAuthentication library handles caching by itself so the user doesn't have to. I don't think it's out of reason to expect the new tech to match that.

cmatskas replied:
Thanks for agreeing to unblock this @schaabs. I know that the team is currently working on a better SQLClient + Azure.identity experience, so good news on this front

schaabs commented on articles/app-service/app-service-web-tutorial-connect-msi.md:
```diff
@@ -136,29 +136,60 @@ Type `Ctrl+F5` to run the app again. The same CRUD app in your browser is now co
+    public override Task<SqlAuthenticationToken> AcquireTokenAsync(SqlAuthenticationParameters parameters)
+    {
+        var tokenRequestContext = new TokenRequestContext(_azureSqlScopes);
+        var tokenResult = _credential.GetToken(tokenRequestContext, default);
```
How often is does AcquireTokenAsync get called? If it's called frequently this code should cache the `AccessToken` returned from `GetToken`. Also since you're acquiring a token in an async context you should call and await `GetTokenAsync` rather than the synchronous `GetToken`.

cmatskas replied:
Updated accordingly. Thanks for the recommendation

cmatskas replied:
@schaabs @cephalin do you have a link that I could point people when adding the comment about best practices and whathaveyou? thx

Commit:
Update articles/app-service/app-service-web-tutorial-connect-msi.md

Co-authored-by: Mickaël Derriey <mderriey@gmail.com>

Commit:
Updated the doc

Provided explanation why we are using `Active Directory Device Code Flow` in the authentication type in the connection string

Commit:
Updated doc with async code

As per the comments, I updated the code sample to use async methods

ktoliver committed:
PR review edit

ktoliver committed:
PR review edit
